### ISSUE TITLE
exported LD_LIBRARY_PATH ignored by RStudio

### ISSUE THREAD

**Issue by milechin** [rstudio/rstudio#12585]
### System details

    RStudio Edition : Desktop
    RStudio Version : 2022.12.0-353
    OS Version      : CentOS 7
    R Version       :  4.2.1

### Steps to reproduce the problem

1. In a bash terminal export `LD_LIBRARY_PATH` with any path.  e.g.  `export LD_LIBRARY_PATH=/path/to/my/lib`
2. Start RStudio and check the value of `LD_LIBRARY_PATH` using R command `Sys.getenv("LD_LIBRARY_PATH")`
3. Compare the value of `LD_LIBRARY_PATH` in RStudio with the one defined in the terminal and see if the path defined in step 1 was included (e.g. in my example it should contain /path/to.my.lib) .

### Describe the problem in detail

I am trying to use an R package called Terra, but it needs a newer version of sqlite3 library.  Since I am working on a cluster that utilizes the module system to manage software and packages, I typically use LD_LIBRARY_PATH to specify which sqlite3 library I want RStudio to use prior to launching RStudio Desktop.  This method worked fine with desktop version 2022.07.2-576, but with RStudio Desktop 2022.12.0-353, the LD_LIBRARY_PATH I set prior to launching RStudio is completely ignored. 

No matter what I set LD_LIBRARY_PATH to in the terminal, it always seems to default to the following.
> Sys.getenv("LD_LIBRARY_PATH")
[1] "/share/pkg.7/r/4.2.1/install/lib64/R/lib:/usr/local/lib64:/usr/java/default/jre/lib/amd64/server"

### Describe the behavior you expected

My expectation is that if I set LD_LIBRARY_PATH in my terminal, that when I launch RStudio Desktop, the path I set in the terminal will be included in the final LD_LIBRARY_PATH that RStudio has set it its environment.

Was there a change in the most recent RStudio Desktop that would cause this new behavior?

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [  ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@milechin Thank you for raising this! I can also reproduce this on RHEL 8. It looks like this has been around since 2022.07, last working in 2022.02. I'll mark this as a bug to review.

**Comment by terjekv** [comment]
I'm also seeing LD_LIBRARY_PATH severely mangled by RStudio 2022.12.0+353 on RHEL9. The linked gist shows the diff between LD_LIBRARY_PATH in the shell that has loaded R as a module, and within RStudio from the same shell. This leads to a broken environment within RStudio as paths such as ` /opt/uio/modules/rhel9/easybuild/software/jbigkit/2.1-GCCcore-11.3.0/lib` are just _gone_. I could try older versions of RStudio, but then I guess I'd have to build them myself as we're on RHEL9.

I did test the nightly, `rstudio-2023.03.0-388-x86_64.rpm`, and it produces the same behaviour as 2022.12.0+353.

The gist with the diff of LD_LIBRARY_PATH: https://gist.github.com/terjekv/dadbfb534e62d23237768ea603cf8357

**Comment by datakid** [comment]
Also seen in customer ticket #88411

**Comment by steklemichael** [comment]
Any updates on this?

**Comment by jthomasmock** [comment]
@ronblum - would you be able to confirm this behavior also affects Workbench/RStudio Pro? My assumption is yes but wanted to confirm for triaging purposes.

**Comment by MariaSemple** [comment]
Just commenting here so I get updates on this ticket.

**Comment by ronblum** [comment]
@jthomasmock Oops, I missed this earlier, sorry for the delay! I tested it in Posit Workbench 2023.03.0+386.pro1, and confirmed that the problem occurs there, too.

**Comment by astrophys** [comment]
I'm experiencing a similar issue on RHEL 8.6 with rstudio-2023.03.0+386 (installed via RPM) in a similar HPC cluster environment.

**Comment by astrophys** [comment]
One way to mitigate this issue is to set `R_ENVIRON` to point to an `Renviron` file and then explicitly set `LD_LIBRARY_PATH`.  It works but is very inflexible 'solution', especially if you need value of from `LD_LIBRARY_PATH` needs to change.

**Comment by astrophys** [comment]
Issue [13100](https://github.com/rstudio/rstudio/issues/13100) appears to be related.

**Comment by astrophys** [comment]
I just installed `rstudio-2023.06.0+421` via the rpm on RHEL 8.6 and it seems to now respect `LD_LIBRARY_PATH`.  It seems to be resolved in the newer version.

**Comment by kevinushey** [comment]
Seems like this might be fixed in the latest daily builds of RStudio? If I run RStudio like so (on Ubuntu 22.04)

```
LD_LIBRARY_PATH=example /usr/bin/rstudio 
```

Then, in RStudio, I see:

```
> Sys.getenv("LD_LIBRARY_PATH")
[1] "/usr/lib/R/lib:/usr/lib/aarch64-linux-gnu:/usr/lib/jvm/default-java/lib/server:example"
```

So we get the library paths computed by R, adjoined with the pre-existing library path.

**Comment by ronblum** [comment]
Tested the following: In `.bashrc`, add the following:
```
export LD_LIBRARY_PATH=/path/to/my/lib:$LD_LIBRARY_PATH
```

The added `/path/to/my/lib` is getting included, but not in a consistent manner. @kevinushey are these accurate?


**Posit Workbench 2023.09.0-daily+293.pro4**
In the Console:
```
> Sys.getenv("LD_LIBRARY_PATH")
[1] "/path/to/my/lib:/opt/R/4.3.1/lib/R/lib:/lib:/usr/local/lib:/usr/lib/jvm/jre-11-openjdk/lib/server"
```

In the Terminal:
```
echo $LD_LIBRARY_PATH
/path/to/my/lib:/path/to/my/lib:/opt/R/4.3.1/lib/R/lib:/lib:/usr/local/lib:/usr/lib/jvm/jre-11-openjdk/lib/server
```

**RStudio Desktop 2023.09.0-daily+301**
In the Console:
```
> Sys.getenv("LD_LIBRARY_PATH")
[1] "/opt/R/4.3.1/lib/R/lib:/usr/local/lib:/usr/lib/jvm/jre-11-openjdk/lib/server:/path/to/my/lib"
```

In the Terminal:
```
[vmware@localhost ~]$ echo $LD_LIBRARY_PATH 
/path/to/my/lib:/opt/R/4.3.1/lib/R/lib:/usr/local/lib:/usr/lib/jvm/jre-11-openjdk/lib/server:/path/to/my/lib
```

**Comment by ronblum** [comment]
@kevinushey or @timtmok The additional `/path/to/my/lib` is included, but in inconsistent locations. Is this acceptable?

**Comment by timtmok** [comment]
I think the inconsistency isn't good. It would also make more sense that the user set paths be at the beginning.

I also couldn't reproduce this issue on Ubuntu 22, Fedora 32, and Fedora 36. Initially, I thought this might be an issue with Centos 7 only.

**Comment by ronblum** [comment]
More info: In RStudio Desktop 2023.09.0-daily+304 on Ubuntu 20.04.5, `LD_LIBRARY_PATH` is added once in the Console and twice in the Terminal.

**RStudio Desktop 2023.09.0-daily+304**
In the Console:
```
> Sys.getenv("LD_LIBRARY_PATH")
[1] "/opt/R/4.2.3/lib/R/lib:/usr/local/lib:/usr/lib/x86_64-linux-gnu:/usr/lib/jvm/java-11-openjdk-amd64/lib/server:/path/to/file:"
```

In the Terminal:
```
$ echo $LD_LIBRARY_PATH 
/path/to/file:/opt/R/4.2.3/lib/R/lib:/usr/local/lib:/usr/lib/x86_64-linux-gnu:/usr/lib/jvm/java-11-openjdk-amd64/lib/server:/path/to/file:
```

**Comment by kevinushey** [comment]
It looks like the base R behavior here is to append the current `LD_LIBRARY_PATH` to whatever it generates:

```
$ LD_LIBRARY_PATH=foo R -s -e 'print(Sys.getenv("LD_LIBRARY_PATH"))'
[1] "/usr/lib/R/lib:/usr/lib/aarch64-linux-gnu:/usr/lib/jvm/default-java/lib/server:foo"
```

So, from that perspective, the desktop behavior is correct and the Workbench behavior is not. However, I'm not sure how comfortable I am changing that around -- it's possible (perhaps even probable?) that some users of workbench are relying on `LD_LIBRARY_PATH` coming first.

For reference, I believe this is where we compute `LD_LIBRARY_PATH` for R sessions:

https://github.com/rstudio/rstudio/blob/9a376ea1029e551f61349956ddca6483544b3ef7/src/cpp/core/r_util/REnvironmentPosix.cpp#L817-L873

If we wanted to change the behavior here, we could move this to the end, after the computation of library paths from the ldpaths script:

https://github.com/rstudio/rstudio/blob/9a376ea1029e551f61349956ddca6483544b3ef7/src/cpp/core/r_util/REnvironmentPosix.cpp#L856-L858

**Comment by jgutman** [comment]
Good catch on the LD_LIBRARY_PATH being added twice. I think this is happening because

1) We add it where Kevin mentions above: 
https://github.com/rstudio/rstudio/blob/9a376ea1029e551f61349956ddca6483544b3ef7/src/cpp/core/r_util/REnvironmentPosix.cpp#L852-L858

2) We also add it here to the `kLibraryPathEnvVariable`, 
https://github.com/rstudio/rstudio/blob/9a376ea1029e551f61349956ddca6483544b3ef7/src/cpp/core/r_util/REnvironmentPosix.cpp#L903-L910

which is getting added as part of default library paths -- so I think it's generally being added twice on Linux Desktop

**Comment by jgutman** [comment]
We also had some original logic about setting and resetting the LD_LIBRARY_PATH in the original Qt code. By a quick scan, it doesn't necessarily look like it was fully duplicated in the electron code, so that plus the original user's mention of this working fine in 2022.07 but breaking in 2022.12 makes me wonder if this is affected by some missed logic in the Electron migration? @ronblum do you think you could test this out?

**Comment by jgutman** [comment]
> So, from that perspective, the desktop behavior is correct and the Workbench behavior is not. However, I'm not sure how comfortable I am changing that around -- it's possible (perhaps even probable?) that some users of workbench are relying on LD_LIBRARY_PATH coming first.

This could explain it, because in Electron, the way we get the LD_LIBRARY_PATH is by running `R --vanilla -s` on `Sys.getenv("${kLdLibraryPathVariable}")`, which returns the post-pended LD LIBRARY PATH, and then we set that result in the process env 
https://github.com/rstudio/rstudio/blob/7a2c7c3ddd30664afacdedbfd9fb4b29fd683abc/src/node/desktop/src/main/detect-r.ts#L194-L204

https://github.com/rstudio/rstudio/blob/7a2c7c3ddd30664afacdedbfd9fb4b29fd683abc/src/node/desktop/src/main/detect-r.ts#L169-L171

So if R is putting this at the end, we will be as well

**Comment by jgutman** [comment]
I've made an update so that a user-defined environment variable (`LD_LIBRARY_PATH` on Linux, `DYLD_FALLBACK_LIBRARY_PATH` on Mac) gets pre-pended to the R system generated value of this variable on Desktop. This is under the assumption that this copies the existing behavior in Workbench/Server, although we'd like to verify this. 

As for the other issues discussed earlier in this thread, I have not been able to reproduce (e.g. `LD_LIBRARY_PATH` getting added to both the beginning and end of the path, user-defined entries being ignored and not reflected in RStudio at all (see [original post](https://github.com/rstudio/rstudio/issues/12585#issue-1536939947) and https://github.com/rstudio/rstudio/issues/13100) , different path outputs in the RStudio console versus the RStudio terminal). 
Let me know if they still persist in the latest dailies and under what conditions

Ideally we'd like to QA that:
- server behavior matches desktop, 
- console output (`Sys.getenv("LD_LIBRARY_PATH")`) matches terminal output (`echo $LD_LIBRARY_PATH`), 
- everything works as expected both when `LD_LIBRARY_PATH` is defined in the environment and when not defined (e.g. RStudio should still produce a value even when the user's shell does not), 
- and this works similarly in Mac as Linux, except using the `DYLD_FALLBACK_LIBRARY_PATH` in Mac (this should have been fixed in #13805)
